### PR TITLE
Add isometric demo

### LINKED ISSUE #1112: examples: Add a basic isometric example
@zacharyburkett will do this :-)

### PR DESCRIPTION
**Pull request by tslocum**
Based on a prototype by Justin Cichra (@jrcichra).

Resolves #1112.

Screenshot

### CODE REVIEW

File: go.mod
Comment (hajimehoshi) [inline go.mod:3]: Do not update this version

File: examples/isometric/game.go
Comment (hajimehoshi) [inline examples/isometric/game.go:1]: 2021

File: examples/isometric/TILES_LICENSE
Comment (hajimehoshi) [inline examples/isometric/TILES_LICENSE:1]: Update examples/resources/images/license.md instead of adding a new file

**Comment by hajimehoshi** [comment]
Thanks!

As I'm preparing for releasing Ebiten v2.2 right now, I'll take a look at your PR and merge this after v2.2 is released. Thank you for understanding.

**Comment by jrcichra** [comment]
This might be a dumb question, but how do I run the example? I tried cloning the fork and running go build:

```
➜  isometric git:(isometric) go build                                            
package github.com/hajimehoshi/ebiten/v2/examples/isometric: build constraints exclude all Go files in /home/justin/git/ebiten-iso/examples/isometric
➜  isometric git:(isometric) pwd          
/home/justin/git/ebiten-iso/examples/isometric
➜  isometric git:(isometric) 
```

Looking at the readme, figured I could run a different example:
```
➜  examples git:(isometric) go run -tags=example github.com/hajimehoshi/ebiten/examples/rotate
no required module provides package github.com/hajimehoshi/ebiten/examples/rotate; to add it:
	go get github.com/hajimehoshi/ebiten/examples/rotate
➜  examples git:(isometric) go get github.com/hajimehoshi/ebiten/examples/rotate
package github.com/hajimehoshi/ebiten/examples/rotate: build constraints exclude all Go files in /home/justin/go/pkg/mod/github.com/hajimehoshi/ebiten@v1.12.12/examples/rotate
➜  examples git:(isometric) 
```

This is with Go 1.17:
```
➜  examples git:(isometric) go version                                          
go version go1.17 linux/amd64
➜  examples git:(isometric)
```

**Comment by hajimehoshi** [comment]
You need to specify a build tag `example`

**Comment by jrcichra** [comment]
Thanks. I was able to run it. This is great!

File: examples/isometric/game.go
Comment (hajimehoshi) [inline examples/isometric/game.go:51]: `2` is fine.

File: examples/isometric/game.go
Comment (hajimehoshi) [inline examples/isometric/game.go:94]: Use `ebiten.IsKeyPressed(key)` instead of `inpututil.KeyPressDurations(key) > 0`.

File: examples/isometric/game.go
Comment (hajimehoshi) [inline examples/isometric/game.go:152]: Do we need to save `op`? What about creating `DrawImageOptions` every time?

File: examples/isometric/game.go
Comment (hajimehoshi) [inline examples/isometric/game.go:168]: We don't need this `if`.

File: examples/isometric/level.go
Comment (hajimehoshi) [inline examples/isometric/level.go:35]: `Tile` for Go convention.

File: examples/isometric/level.go
Comment (hajimehoshi) [inline examples/isometric/level.go:43]: `Size` for Go convention.

File: examples/isometric/level.go
Comment (hajimehoshi) [inline examples/isometric/level.go:72]: Can we use integers for the random number like `r.Intn(1000)`?

File: examples/isometric/main.go
Comment (hajimehoshi) [inline examples/isometric/main.go:21]: Why is this needed here?

File: examples/isometric/tile.go
Comment (hajimehoshi) [inline examples/isometric/tile.go:37]: This can cause a resource leak, as all the elements in the underlying array are still held. Let's clear the elements explicitly.

```
for i := range t.sprites {
    t.sprites[i] = nil
}
t.sprites = t.sprites[:0]
```

File: examples/isometric/game.go
Comment (tslocum) [reply examples/isometric/game.go:152]: Wouldn't that introduce an allocation each `Draw`?

File: examples/isometric/main.go
Comment (tslocum) [reply examples/isometric/main.go:21]: Without this, we get the following error:

`failed to load embedded spritesheet: image: unknown format`

Comment (hajimehoshi) [reply examples/isometric/main.go:21]: This should be moved to where decoding actually happens.

File: examples/isometric/game.go
Comment (hajimehoshi) [reply examples/isometric/game.go:152]: Creating a DrawImageOption every frame should not be so problematic (from my experience).

File: examples/isometric/game.go
Comment (hajimehoshi) [inline examples/isometric/game.go:207]: Create a local variable `op` outside of the loop and use it.

File: examples/isometric/game.go
Comment (hajimehoshi) [inline examples/isometric/game.go:103]: 0 seems a valid value as CursorPosition. What about using `math.MinInt32` indicating an invalid value?

File: examples/isometric/game.go
Comment (hajimehoshi) [inline examples/isometric/game.go:101]: What is 'pan'?

File: examples/isometric/game.go
Comment (hajimehoshi) [inline examples/isometric/game.go:147]: FilterNearest is the default value and you don't have to specify this.

File: examples/isometric/game.go
Comment (hajimehoshi) [inline examples/isometric/game.go:175]: This function is not used anywhere.

File: examples/isometric/game.go
Comment (hajimehoshi) [inline examples/isometric/game.go:189]: I slightly prefer the loop for y outside and the loop for x inside. Would it be possible to write like this?

```
for y := 0; y < g.currentLevel.h; y++ {
    // ...
   for x := 0; x < g.currentLevel.w; x++ {
```

If this is not easy, the current code is fine.

File: examples/isometric/level.go
Comment (hajimehoshi) [inline examples/isometric/level.go:59]: NewLevel should return an error instead of logging here.

File: examples/isometric/level.go
Comment (hajimehoshi) [inline examples/isometric/level.go:69]: ditto

File: examples/isometric/main.go
Comment (hajimehoshi) [inline examples/isometric/main.go:30]: Is 144 intended? If so, why?

File: examples/isometric/game.go
Comment (hajimehoshi) [reply examples/isometric/game.go:103]: Or

```
	maxInt     = int(^uint(0) >> 1)
	minInt     = -maxInt - 1
```

and use `minInt`

File: examples/isometric/main.go
Comment (tslocum) [reply examples/isometric/main.go:30]: This ensures 144Hz displays are satisfied, as lower TPS can cause duplicated frames and higher TPS would not result in any perceptual difference on higher Hz displays.

File: examples/isometric/game.go
Comment (tslocum) [reply examples/isometric/game.go:101]: >To pivot a movie camera along a horizontal plane in order to follow an object or create a panoramic effect.

File: examples/isometric/game.go
Comment (hajimehoshi) [reply examples/isometric/game.go:187]: As an example, I don't want to have a function that is not used anywhere.

File: examples/isometric/main.go
Comment (hajimehoshi) [reply examples/isometric/main.go:30]: I don't think this example is so FPS sensitive that this has to have 144 Hz. For consistency with other examples, would it be possible to remove this? Thanks,

File: examples/isometric/game.go
Comment (tslocum) [reply examples/isometric/game.go]: Including this function makes the example more complete IMO as it is necessary to support interaction (namely mouse).  Anyone who wants to actually make something using this demo as a starting point will need it.  Nevertheless, I've removed it.

Comment (hajimehoshi) [reply examples/isometric/game.go]: I understand this function might be useful for users. If you want to leave this, I'm fine to leave this as a comment here.

**Review by hajimehoshi — APPROVED** [review]
LGTM, thanks!

I'm waiting for your reply about the comment (whether you want to leave a comment `isoToCartesian`

File: examples/isometric/game.go
Comment (tslocum) [reply examples/isometric/game.go]: Are you asking if I'd like to document this function by adding it to a PR review comment?  Sorry if I am misunderstanding.

If that is what you meant, I believe the function is already documented in this way as it is attached to this comment thread above.  Thank you ,though.

Comment (tslocum) [reply examples/isometric/game.go:188]: Derp, I just realized what you meant. Including this function within a comment block sounds great.  I've updated the PR.

Comment (hajimehoshi) [reply examples/isometric/game.go:188]: I think more comments are needed to explain, like, "this function might be useful for those who want to extend this example".

File: examples/isometric/main.go
Comment (hajimehoshi) [inline examples/isometric/main.go:30]: Wait, why isn't this vsynced? Tearing happened on my machine.

File: examples/isometric/game.go
Comment (hajimehoshi) [inline examples/isometric/game.go:158]: Why do we need this spinner animation?

**Comment by tslocum** [comment]
I am a little frustrated with the way this PR review is unfolding.  We are about to start a fourth round of rested changes.  In the previous three rounds, I have resolved each requested  change.  Please let me know when you are finished requesting changes, as I don't want to waste any more of my time doing this in peace-meal fashion.

**Comment by hajimehoshi** [comment]
I'm afraid this is usual especially for a new example from scratch (e.g. https://github.com/hajimehoshi/ebiten/pull/1276). I'm sorry I didn't realized the last two points unless I actually executed the PR. I think these are last requests.

File: examples/isometric/game.go
Comment (hajimehoshi) [inline examples/isometric/game.go:30]: This is no longer needed, but I'll remove this.

**Review by hajimehoshi — APPROVED** [review]
Thank you!

**Pull request merged**
